File: data-structure-and-algorithm/java/set-matrix-zeroes.java
/**
 * Set Matrix Zeroes
 * 
 * Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in place.
 * 
 * click to show follow up.
 * 
 * Follow up:
 * Did you use extra space?
 * A straight forward solution using O(mn) space is probably a bad idea.
 * A simple improvement uses O(m + n) space, but still not the best solution.
 * Could you devise a constant space solution?
 */

public class Solution {
    public void setZeroes(int[][] matrix) {
        if (matrix == null) {
            throw new NullPointerException();
        }

        if (matrix.length == 0 || matrix[0].length == 0) {
            return;
        }

        // determine zero exists in first row, and first column
        boolean doesFirstColHaveZero = false, doesFirstRowHaveZero = false;
        for (int i = 0; i < matrix.length; ++i) {
            if (matrix[i][0] == 0) {
                doesFirstColHaveZero = true;
                break;
            }
        }

        for (int j = 0; j < matrix[0].length; ++j) {
            if (matrix[0][j] == 0) {
                doesFirstRowHaveZero = true;
                break;
            }
        }

        // shadow zero onto first row, and column from rows, and columns starting second which have zero
        for (int i = 1; i < matrix.length; ++i) {
            for (int j = 1; j < matrix[0].length; ++j) {
                if (matrix[i][j] == 0) {
                    matrix[0][j] = 0;
                    matrix[i][0] = 0;
                }
            }
        }

        // fill zeroes onto rows and columns starting second, based on shadowing
        for (int i = 1; i < matrix.length; ++i) {
            for (int j = 1; j < matrix[0].length; ++j) {
                if (matrix[0][j] == 0 || matrix[i][0] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }

        // fill first row, and column if they originally have zero
        if (doesFirstColHaveZero) {
            for (int i = 0; i < matrix.length; ++i) {
                matrix[i][0] = 0;
            }
        }

        if (doesFirstRowHaveZero) {
            for (int j = 0; j < matrix[0].length; ++j) {
                matrix[0][j] = 0;
            }
        }
    }
}

File: data-structure-and-algorithm/cpp/set-matrix-zeroes.cpp
// 73. Set Matrix Zeroes
// Difficulty: Medium

// Given a m x n matrix, if an element is 0, set its entire row and column to 0. Do it in place.

// Follow up:
// Did you use extra space?
// A straight forward solution using O(mn) space is probably a bad idea.
// A simple improvement uses O(m + n) space, but still not the best solution.
// Could you devise a constant space solution?

// Time:  O(m * n)
// Space: O(1)

class Solution {
public:
  void setZeroes(vector<vector<int>>& matrix) {
    bool hasZero = false;
    int firstRowZero = -1, firstColZero = -1;

    for (int i = 0; i < matrix.size(); ++i) {
      for (int j = 0; j < matrix[0].size(); ++j) {
        if (matrix[i][j] == 0) {
          if (!hasZero) {
            firstRowZero = i;
            firstColZero = j;
            hasZero = true;
          }
          matrix[firstRowZero][j] = 0;
          matrix[i][firstColZero] = 0;
        }
      }
    }

    if (hasZero) {
      for (int i = 0; i < matrix.size(); ++i) {
        if (i == firstRowZero) {
          continue;
        }
        for (int j = 0; j < matrix[0].size(); ++j) {
          if (j == firstColZero) {
            continue;
          }
          if (matrix[firstRowZero][j] == 0 || matrix[i][firstColZero] == 0) {
            matrix[i][j] = 0;
          }
        }
      }
      // process 1st found-zero col
      for (int i = 0; i < matrix.size(); ++i) {
        matrix[i][firstColZero] = 0;
      }
      // process 1st found-zero row
      for (int j = 0; j < matrix[0].size(); ++j) {
        matrix[firstRowZero][j] = 0;
      }
    }
  }
};
